sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala

```diff
@@ -809,6 +860,10 @@ case class SessionWindowStateStoreSaveExec(
       newNumRowsUpdated = stateOpProgress.numRowsUpdated,
       newNumRowsDroppedByWatermark = numRowsDroppedByWatermark)
   }
+
+  // This operator will evict based on min input watermark and ensure it will be minimum of
+  // the event time value for the output so far (including output from eviction).
+  override def produceWatermark(inputWatermarkMs: Long): Long = inputWatermarkMs
```

Comment:
How about making this the default implementation so that no all stateful operators need to define this default behavior.

Reply:
Yeah... mostly this method will be only implemented via built-in operators and most operators don't need to adjust the watermark. Sounds good.